File: QEasyWidgets/Common/Translator.py
from pathlib import Path
from typing import Union, Optional
from PySide6.QtCore import QTranslator
from PySide6.QtWidgets import QApplication

from .Signals import ComponentsSignals
from .Language import EasyLanguage, currentLanguage

##############################################################################################################################

class TranslationBase(QTranslator):
    '''
    '''
    def __init__(self, parent = None):
        super().__init__(parent)

    def load(self, language: Optional[str] = None):
        EasyLanguage.update(language) if language is not None else None

        Prefix = 'QM'
        FilePath = f'i18n/{currentLanguage()}.qm'
        FilePath = Path(f':/{Prefix}').joinpath(FilePath).as_posix()

        super().load(FilePath)


Translator = None
def Function_UpdateLanguage(
    language: Optional[str] = None
):
    '''
    '''
    global Translator

    QApplication.instance().processEvents()

    # Remove old translator
    QApplication.instance().removeTranslator(Translator) if Translator is not None else None

    # Install new translator
    Translator = TranslationBase()
    Translator.load(language)
    QApplication.instance().installTranslator(Translator)


ComponentsSignals.Signal_SetLanguage.connect(Function_UpdateLanguage)

##############################################################################################################################

File: c++/QEasyWidgets/Common/Translator.cpp
#include "Translator.h"

#include <QDir>
#include <QCoreApplication>


TranslationBase::TranslationBase(QObject *parent)
    : QTranslator(parent) {
}


bool TranslationBase::load(Language language) {
    QString filePath = getTranslationFilePath(language);
    return QTranslator::load(filePath);
}


bool TranslationBase::load(const QString &languageStr) {
    Language lang = languageStr.isEmpty() ? currentLanguage() : stringToLanguage(languageStr);
    return load(lang);
}


QString TranslationBase::getTranslationFilePath(Language language) const {
    QString langStr = languageString(language);
    QString filePath = QString(":/QM/i18n/%1.qm").arg(langStr);
    return filePath;
}


TranslationBase *translator = nullptr;


void updateLanguage(Language language) {
    if (!QApplication::instance()) {
        return;
    }

    QApplication::instance()->processEvents();

    // Remove old translator
    if (translator) {
        QApplication::instance()->removeTranslator(translator);
        delete translator;
        translator = nullptr;
    }

    // Install new translator
    translator = new TranslationBase();
    if (translator->load(language)) {
        QApplication::instance()->installTranslator(translator);
    } else {
        delete translator;
        translator = nullptr;
    }

    EasyLanguage.setLanguage(language);
}


void updateLanguage(const QString &languageStr) {
    Language lang = languageStr.isEmpty() ? currentLanguage() : stringToLanguage(languageStr);
    updateLanguage(lang);
}